GH-49102: [CI] Add type checking infrastructure and CI workflow for type annotations

rok opened the pull request:
### Rationale for this change

This is the first in series of PRs adding type annotations to pyarrow and resolving #32609.

### What changes are included in this PR?

This PR establishes infrastructure for type checking:

- Adds CI workflow for running mypy, pyright, and ty type checkers on linux, macos and windows
- Configures type checkers to validate stub files (excluding source files for now)
- Adds PEP 561 `py.typed` marker to enable type checking
- Updates wheel build scripts to include stub files in distributions
- Creates initial minimal stub directory structure
- Updates developer documentation with type checking workflow

### Are these changes tested?

No. This is mostly a CI change.

### Are there any user-facing changes?

This does not add any actual annotations (only `py.typed` marker) so user should not be affected.
* GitHub Issue: #32609
* GitHub Issue: #49102

rok committed:
Add GitHub Actions workflow for type checking (mypy, pyright, ty). Add script for including docstrings into stubfiles before building wheels.

AlenkaF reviewed (approved):
Thank you @rok for working on this! Splitting up the main PR sounds like a great idea to me.

I only have one small question, otherwise the changes look good to me.

AlenkaF commented:
Yes, let's wait for Raul 👍

rok commented on ci/scripts/python_test_type_annotations.sh:
```diff
@@ -0,0 +1,40 @@
+
+if [ "${PYARROW_TEST_ANNOTATIONS}" == "ON" ]; then
+  # Install library stubs
+  pip install pandas-stubs scipy-stubs sphinx types-cffi types-psutil types-requests types-python-dateutil
```
If I remove sphinx I get the following on the final PR:
```
pyarrow/vendored/docscrape.py:622: error: Cannot find implementation or library stub for module named "sphinx.ext.autodoc"  [import-not-found]
pyarrow/vendored/docscrape.py:622: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
```

So sphinx is not needed for functioning of pyarrow and it's probably better to then inline ignore this error (once we enable typechecking). Will remove sphinx here.

rok replied:
```suggestion
  pip install pandas-stubs scipy-stubs types-cffi types-psutil types-requests types-python-dateutil
```

rok committed:
Update ci/scripts/python_test_type_annotations.sh

raulcd commented on ci/scripts/python_test_type_annotations.sh:
```diff
@@ -0,0 +1,40 @@
+
+set -ex
+pyarrow_dir=${1}
+
```
Can we add the virtualenv activation in case of `ARROW_PYTHON_VENV` present? This is a pattern we use on lots of our sh scripts.
```
if [ -n "${ARROW_PYTHON_VENV:-}" ]; then
  . "${ARROW_PYTHON_VENV}/bin/activate"
fi
```

raulcd commented on ci/scripts/python_wheel_validate_contents.py:
```diff
@@ -35,6 +35,11 @@ def validate_wheel(path):
     assert not outliers, f"Unexpected contents in wheel: {sorted(outliers)}"
     print(f"The wheel: {wheels[0]} seems valid.")
 
+    candidates = [info for info in f.filelist if info.filename.endswith('compute.pyi')]
```
Won't this fail also? I don't think we are able to generate those, see the other comment about the missing `dev/update_stub_docstrings.py`
```suggestion
    # Validate at least one typing stub has been generated.
    candidates = [info for info in f.filelist if info.filename.endswith('compute.pyi')]
```

rok replied:
Replaced with TODO.

raulcd commented on ci/scripts/python_wheel_xlinux_build.sh:
```diff
@@ -167,6 +167,11 @@ export ARROW_HOME=/tmp/arrow-dist
+# We first populate stub docstrings and then build the wheel
+python setup.py build_ext --inplace
+python -m pip install griffe libcst
+python ../dev/update_stub_docstrings.py pyarrow-stubs
```
this will fail, right? this file isn't present on the current PR so our wheels will start failing to build if we merge as is? I'll run archery just to validate

raulcd reviewed (commented):
Thanks for the PR @rok sorry it took me a while to review with holidays, release, etcetera

rok committed:
Apply suggestion from @raulcd

Co-authored-by: Raúl Cumplido <raulcumplido@gmail.com>

rok committed:
review feedback

rok commented on compose.yaml:
```diff
@@ -1001,14 +1003,16 @@ services:
       ARROW_S3: "OFF"
       ARROW_SUBSTRAIT: "OFF"
       ARROW_WITH_OPENTELEMETRY: "OFF"
+      PYARROW_TEST_ANNOTATIONS: "ON"
```
@raulcd do you think we actually want to be able to toggle these? I'm thinking we don't really need them

rok committed:
include dev/update_stub_docstrings.py

rok committed:
Use PYARROW_TEST_ANNOTATIONS in windows build, disable wheel docstring check

rok committed:
work on dev/update_stub_docstrings.py

rok committed:
further work on dev/update_stub_docstrings.py

rok committed:
lint

rok committed:
add click for docstring population

rok commented:
@github-actions crossbow submit wheel-manylinux-2-28-cp311-cp311-amd64

rok committed:
Remove dependencies on click and griffe

rok committed:
fix import paths

rok committed:
add PYARROW_TEST_ANNOTATIONS to AMD64 Windows 2022 Python 3.13
change bat
lint
add a popd and nicer logging for windows
ReplaceElipsis -> DocstringInserter
simplify
remove sphinx

rok committed:
move check to pre-commit

rok committed:
change pre-commit, add note

rok committed:
change pre-commit script

rok committed:
Revert from pre-commit to ci/scripts/python_test_type_annotations.sh

kou reviewed (approved):
+1

rok committed:
Apply suggestions from code review

Co-authored-by: Sutou Kouhei <kou@cozmixng.org>

rok committed:
apply review suggestion

rok committed:
fix shellcheck

rok committed:
Try single build

rok committed:
lint

pitrou commented on ci/scripts/python_test_type_annotations.sh:
```diff
@@ -0,0 +1,43 @@
+set -ex
+pyarrow_dir=${1}
+
+if [ "${PYARROW_TEST_ANNOTATIONS}" == "ON" ]; then
```
Why this? Presumably one should only call this script if they want to actually check annotations?

rok replied:
Removing here.

rok replied:
Actually let's get @raulcd's opinion in light of our previous discussion: https://github.com/apache/arrow/pull/48618#discussion_r2690100864

raulcd replied:
I agree with @pitrou my previous comment was that the check is necessary if it gets integrated on the `python_test.sh` script otherwise, if we use a different script as we are currently doing, we should remove it.

pitrou reviewed (commented):
The general idea seems fine to me. Some early comments below. I think @raulcd might want to take a closer look at the build setup.

rok committed:
review feedback

rok commented on ci/scripts/python_wheel_macos_build.sh:
```diff
@@ -177,6 +177,8 @@ export CMAKE_PREFIX_PATH=${build_dir}/install
 
 pushd ${source_dir}/python
+# Install libcst for build-time stub docstring extraction
+python -m pip install libcst
```
Done, see if it makes sense.

rok commented on ci/scripts/python_wheel_validate_contents.py:
```diff
@@ -34,7 +34,7 @@ def validate_wheel(path):
     assert not outliers, f"Unexpected contents in wheel: {sorted(outliers)}"
     print(f"The wheel: {wheels[0]} seems valid.")
-
+    # TODO(GH-32609): Validate some docstrings were generated and added.
```
Will do.

rok replied:
I created two issues (for this PR and one following it) and linked them into the original issue https://github.com/apache/arrow/issues/32609

rok commented on dev/update_stub_docstrings.py:
```diff
@@ -0,0 +1,233 @@
+    parser.add_argument("stubs_dir", type=Path, help="Path to pyarrow-stubs folder")
+    args = parser.parse_args()
+
+    sys.path.insert(0, ".")
```
Using `setup_dir` now.

rok committed:
reintroduce pythonVersion = 3.10

rok commented on python/pyproject.toml:
```diff
@@ -84,15 +86,51 @@ zip-safe=false
+# TODO: Enable type checking once stubs are merged
+[tool.pyright]
+pythonPlatform = "All"
+pythonVersion = "3.10"
```
Maybe we best keep it in then?

rok commented on python/pyproject.toml:
```diff
@@ -18,6 +18,8 @@
 requires = [
     "cython >= 3.1",
+    # Needed for build-time stub docstring extraction
+    "libcst>=1.8.6",
```
That is probably best done in another PR?

rok committed:
revert change

rok committed:
post rebase fix

rok commented:
Thanks for the review @pitrou. I addressed your comments and rebases. Please do another pass.

rok committed:
--no-build-isolation bypasses pyproject.toml's build requirements

rok committed:
libcs added

rok committed:
Added DYLD_LIBRARY_PATH export to python_build.sh

rok committed:
Remove some sys.path hackery

rok committed:
trigger CI

robsdedude commented on python/pyarrow/py.typed:
I think adding a `py.typed` marker to the package before all public APIs are typed (even just with a whole bunch of `Any`s) wasn't ideal. Downstream users will now see a bunch of typing errors because type checkers now assume that pyarrow provides type hints that can be relied upon.

In my case, I'm seeing `mypy` complain about `Module has no attribute "count"  [attr-defined]` where the module in question is `pyarrow.compute`.

Please consider either removing the marker until type hints are ready or to make sure all public APIs exist (at least as `def xyz(*args: Any, **kwargs: Any) -> Any` stubs). Or apply the same `__getatrr__` placeholder to all modules, not just the top-level module.

rok replied:
Agreed that this is not ideal  @robsdedude. Unfortunately progress on annotations was slow. I'll try to get more coverage merged by next release and look into `__getatrr__` as a temporary measure. Work continues at https://github.com/apache/arrow/pull/48622, reviews welcome :)

robsdedude replied:
For reference (created this before you comment for better visibility): https://github.com/apache/arrow/issues/50123
